### ISSUE TITLE
`useRefocus` hook does not trigger

### ISSUE THREAD

**Issue by tofumatt** [google/site-kit-wp#8297]
## Bug Description

As noticed by @aaemnnosttv and I: the `useRefocus` hook does not actually trigger when a user navigates away from the tab or browser window entirely when testing #8294: https://10up.slack.com/archives/CBKKQEBR9/p1708531310258029

It seems there's a bug in the hook's behaviour. It might have to do with Chrome's throttling of background `setTimeout` calls, but probably not, given in my testing it is _never_ triggered, not simply delayed.

## Steps to reproduce

First, add a `console.log` statement to the `onFocus` function, eg.:
![Screenshot 2024-02-21 at 19 18 01](https://github.com/google/site-kit-wp/assets/90871/df315c3f-4acb-4254-8b9c-6eed27221c40)

1. Set up Site Kit and connect Analytics
2. Go to set up AdSense, so you trigger the `SetupMain` component.
3. Navigate away from the tab/window
4. Wait 16 seconds
5. Navigate back to the tab/window

Notice the console statement is never output.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The `useRefocus` hook should be triggered when the user restores "focus" to the tab that the hook was called from.

## Implementation Brief

Note: Please refer this [comment](https://github.com/google/site-kit-wp/issues/8297#issuecomment-4460058562) on why this issue is happening at first place.

- [x] In `assets/js/hooks/useRefocus.js`
  - Use `useRef` hook to get `callbackRef` for `callback` function passed to the hook.
  - Use the `useEffect` hook to store callback in `callbackRef`
  - Call the callback function via ref instead of calling `callback` directly, so that callback reference is not changed when state is updated for the component.
  - Remove `callback` from the dependency array of cleanup function.

### Test Coverage

- Add a test to simulate callback prop change after blue event, verify still fires on focus.
- Existing tests should keep passing for the `useRefocus` hook.

## QA:Eng Brief

1. Add the `UseRefocusBugReproduction` [component](https://github.com/google/site-kit-wp/blob/70e772c061660c49f078352d659efecb83c35ce8/assets/js/hooks/useRefocus-bug-reproduction.js) inside `assets/js/hooks`.

2. Add to Story with the reproduction component, you can get it from [here](https://github.com/google/site-kit-wp/blob/70e772c061660c49f078352d659efecb83c35ce8/assets/js/hooks/useRefocus-bug-reproduction.stories.js):

   ```javascript
   import UseRefocusBugReproduction from "@/js/hooks/useRefocus-bug-reproduction";

   export default function TestPage() {
     return <UseRefocusBugReproduction />;
   }
   ```

3. **Open the page in browser and open DevTools console**

#### Steps to See the Bug:

1. **Click "🔴 Leave Tab (Blur)" button**
   - Simulates blur event
   - Console shows: `[useRefocus] Blur event - waiting 5000ms before sync`
   - A 5-second timer starts

2. **Within 5 seconds, click "⚠️ Toggle State (Trigger Bug)" button**
   - This changes internal state
   - New callback reference is created (due to useCallback dependency)
   - useRefocus detects new callback in dependency array
   - Console shows: `[useRefocus] useEffect cleanup - clearing timeout` ← **The bug!**
   - The 5-second timeout set in step 1 is now **cleared prematurely**

3. **Click "🟢 Return to Tab (Focus)" button** (doesn't matter if 5 seconds have passed)
   - Simulates focus event
   - Console shows: `[useRefocus] Focus event - but timeout was cleared or not yet elapsed`
   - **Expected:** Sync Execution Count should increment
   - **Actual:** Counter doesn't change because callback never ran (BUG!)

#### Console Log Pattern (Bug Visible on main branch):

```
[useRefocus] Blur event - waiting 5000ms before sync
[useRefocus-Bug-Repro] State toggled - new callback reference will be created
[useRefocus] useEffect cleanup - clearing timeout              ← BUG: Timeout cleared!
[useRefocus] Focus event - but timeout was cleared...         ← Callback doesn't execute
(Missing: [useRefocus] onSync callback executed!)             ← BUG CONFIRMED!
```

---

## Expected Behavior (After Fix on develop branch after merge)

After the fix is applied, the pattern should be:

```
[useRefocus] Blur event - waiting 5000ms before sync
[useRefocus-Bug-Repro] State toggled - new callback reference will be created
(NO cleanup message - effect not re-initialized unnecessarily)
(Wait for 5 seconds OR click focus earlier)
[useRefocus] Timeout elapsed after 5000ms
[useRefocus] Focus event - executing callback
[useRefocus] onSync callback executed!                        ← SUCCESS!
```

## Changelog entry

* Ensure that the `useRefocus()` hook's callback fires consistently in all cases.

**Comment by derweili** [comment]
I followed @tofumatt steps but can't replicate. I added some more console logs to investigate the behavior in more detail, but `useRefocus()` hook was always working for me as expected.

I tried following `blur` scenarios:
- switching tab
- switching window
- minimizing window and switch to a different app

The console logs were always triggered correctly.

However there is one behavior that was unexpected for me: When you switch back to the tab while dev tools are open, the `focus` event isn't called. This is because the focus is on the devtools, not on the window. You have to click on the main window first for the `focus` event to be triggered.

>It might have to do with Chrome's throttling of background setTimeout calls

While I wasn't able to replicate this, we might want to switch to storing a timestamp on `blur` instead of relying on a timeout.

**Comment by benbowler** [comment]
Similar to @derweili I wasn't able to repeat the original issue, although I couldn't get Chrome to start tab freezing as I struggled to max out my 64Gb of RAM.

**Comment by binnieshah** [comment]
@aaemnnosttv - @zutigrm and @benbowler have confirmed they are unable to reproduce this issue. I have removed the next up label, but am wondering is this still an issue? If not can we close?

**Comment by ankitrox** [comment]
@binnieshah I've spent some time trying to reproduce the issue before we close this one, but I am not able to recreate this issue and `useRefocus` hook is working as expected.

<img width="1382" alt="Screenshot 2024-07-11 at 6 29 02 PM" src="https://github.com/google/site-kit-wp/assets/1442637/dfe3276d-afd8-4ada-9f36-c2ad674ae9b2">


I am getting the expected output when I come back to the tab from the other one or re-open the window.

The dev tools related note added by @derweili is valid, but that's not the issue with hook.

To sum up, everything seems to be working as expected. I believe, if everyone here agrees, we can close this one.

Thank you.

**Comment by ankitrox** [comment]
Unassigning myself from the issue as there are no steps required further.

**Comment by aaemnnosttv** [comment]
@tofumatt assigning to you to review. Is this still an issue? If not, you know what to do 😄

**Comment by techanvil** [comment]
Hey folks, just a heads up that we've seen some similar behaviour while implementing #9262, and it looks like the cleanup handler for the `useEffect()` in `useRefocus()` is triggering unexpected when clicking away from the tab. Maybe a similar thing is happening here (if it's still reproducible in this scenario). See https://github.com/google/site-kit-wp/pull/9356/files#r1763034113 for more info and a screencast.

**Comment by techanvil** [comment]
Update: I did find a fix for the issue in #9262, which may have some relevance here: see https://github.com/google/site-kit-wp/pull/9356#discussion_r1763452720 for more detail on this.

**Comment by ankitrox** [comment]
### Root cause of the issue

- `useRefocus` includes `callback` in the dependency array of the cleanup function.
- When a parent component re-renders with a new callback reference,(which happens after state updates), the `useEffect` cleanup runs.
- The cleanup clears the timeout that was set by the recent blur event
- When the user refocuses, the callback never runs because the timeout was cleared before it could elapse.

**Comment by techanvil** [comment]
IB ✅

**Comment by JakePT** [comment]
✅ QA:Eng
